test/unit/math/prim/core/set_num_threads_test.cpp

```diff
@@ -0,0 +1,10 @@
+#include <stan/math/prim/core.hpp>
+#include <gtest/gtest.h>
+
+TEST(intel_tbb_new_init, set_threads) {
+  int num_threads = std::thread::hardware_concurrency();
+  auto& tbb_init = stan::math::init_threadpool_tbb(num_threads);
+  EXPECT_TRUE(tbb_init.is_active());
+
+  EXPECT_EQ(num_threads, tbb_init.max_concurrency());
+}
```

Comment:
Weird Jenkins testing thing, I tested this locally with 
```cpp
int num_threads = std::thread::hardware_concurrency() - 1;
```
but it won't work on Jenkins because we merge all these tests together so by the time this tests run the max threadpool has already been set so setting it lower doesn't do anything. imo I think it's find unless theres some sort of quick fix

Reply:
@SteveBronder Try to run the tests inside the initialized arena, using `tbb_arena.execute([&]() {...})` like:
https://github.com/stan-dev/math/blob/818be4ae86ae91af53943dc742ce977102bd1689/test/unit/math/prim/core/init_threadpool_tbb_late_test.cpp#L17-L24

or terminate the arena after each test, using `tbb_arena.terminate()`.

Reply:
Or exclude prim/core from jumbo here: https://github.com/stan-dev/math/blob/develop/runTests.py#L35

Reply:
@hsbadr I couldn't figure out how to do the above. Even if you terminate the task arena are you able to then later change the number of threads the global task arena has?

@rok-cesnovar I went with your suggestion as it's a bit easier and we only start this up once at runtime so it makes sense to test this individually

Reply:
Sounds good. I don't know how important the excluded `prim/core` tests are, but it's indeed the easiest workaround.

`stan::math::init_threadpool_tbb()` creates static instances of `global_control` and `arena`, which restores previous parallelism limitation on exit. Also, `arena.execute()` should always work as expected, and it's the best solution if we'd like to run Stan services in separate TBB arenas, with custom `num_threads`, for example.

Reply:
ftr they are not excluded they are just not put together into one big file and instead each run as separate test programs.

> stan::math::init_threadpool_tbb() creates static instances of global_control and arena, which restores previous parallelism limitation on exit. Also, arena.execute() should always work as expected, and it's the best solution if we'd like to run Stan services in separate TBB arenas, with custom num_threads, for example.

I thought I was doing it right but putting them all together into one file I wasn't able to change the number of threads. Maybe I just read the docs wrong?

Reply:
> ftr they are not excluded they are just not put together into one big file and instead each run as separate test programs.

Ah! I see.

> I thought I was doing it right but putting them all together into one file I wasn't able to change the number of threads. Maybe I just read the docs wrong?

I think we can't reset the parallelism limitation (`tbb::global_control::max_allowed_parallelism`, set statically by `stan::math::init_threadpool_tbb()`) until we exit from main (or the function that created it). Alternatively, we can terminate the arena and initiate a new one. This is typically done by setting a `global_control` that limits the maximum parallelism and schedule/subset the number of threads by multiple arenas (executing code inside each arena via `arena.execute()`) with `num_threads <= bb::global_control::max_allowed_parallelism`.

This is a lot of work though, as you need to edit all the code that run in parallel.